### PR TITLE
Phillip/my courses added basic subscribe/unsubscribe functions

### PR DESCRIPTION
**Pull request by phiphi282**
This PR contains:

- an endpoint that returns all subscribed users for a course

- an endpoint that returns all courses a user is subscribed to

- an endpoint to subscribe to a course

- an endpoint to unsubscribe from a course

this pr still need testing, but i still opened it for better coverage overseeing.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -39,7 +39,34 @@ You can start it over the command line (your container name might be different f
 ### User
 - `/api/v1/me` GET own user (based on SessionToken).
 - `/api/v1/users/:id` GET users.
+- `/api/v1/users/:id/courses` GET all courses a member is subscribed to
```
Comment (ldb) [inline README.md]: 👍 Could you also add the other endpoints that were added in the meanwhile? I totally forgot about this.

File: course.go [hunk 2]
```diff
@@ -32,6 +32,18 @@ type CourseUpdater interface {
 	Update(id string, update bson.M) (error, Course)
 }
 
+type CoursesByMemberFinder interface {
```
Comment (ldb) [inline course.go]: I think we should keep the repository interfaces as small as possible. Recently I removed a lot of functions from there and replaced them with the generic ones that should be powerful enough to do everything needed:
`FindOne`, `FindMany`, `Update`, `Delete` and `Insert`. Those three functions could all be implemented by using `FindMany` and `Update`.

File: http/courseHandler.go [hunk 3]
```diff
@@ -85,3 +85,91 @@ func (a *AppServer) GetCourseHandler() httprouter.Handle {
...
+			a.Logger.Printf("Error while subscribing user to course %v", err)
+			w.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+
+	}
+}
+
+func (a *AppServer) RemoveMembersHandler() httprouter.Handle {
+	type request struct {
```
Comment (ldb) [inline http/courseHandler.go:151]: Does this actually work? In the readme you defined the request as JSON like this : `[{id},{id}]`. Wouldn't that be a `type request []struct...`? Speaking of which, you could as well define it like this: `{"members":[id, id]}`. Might be a bit clearer.

File: http/courseHandler.go [hunk 4]
```diff
@@ -85,3 +85,91 @@ func (a *AppServer) GetCourseHandler() httprouter.Handle {
...
+		ID string `json:"id"`
+	}
+	return func(w http.ResponseWriter, r *http.Request, p httprouter.Params) {
+		request := []request{}
```
Comment (ldb) [inline http/courseHandler.go:155]: Ah, I see now you did it here.

File: http/courseHandler.go [hunk 5]
```diff
@@ -85,3 +85,91 @@ func (a *AppServer) GetCourseHandler() httprouter.Handle {
...
+			a.Logger.Printf("Error decoding body %v", err)
+			w.WriteHeader(http.StatusBadRequest)
+			return
+		}
+		members := []string{}
+		for _, req := range request {
+			members = append(members, req.ID)
+		}
+		err, _ = a.CourseService.RemoveMembers(id, members)
+
```
Comment (ldb) [inline http/courseHandler.go]: Newline.

File: http/courseHandler_test.go [hunk 6]
```diff
@@ -110,7 +111,131 @@ func TestAppServer_GetCourseHandler(t *testing.T) {
...
+	}{
+		{"success", "1", `[{"id": "1"},{"id": "2"}]`, 200},
+		{"invalid body", "2", `lala <> ""`, 400},
+		{"error subscribing user", "3", `[{"id": "1"},{"id": "2"}]`, 500},
+	}
+
+	mockService.RemoveMembersFn = func(id string, members []string) (error, eduboard.Course) {
+		switch id {
+		case "1":
+			if len(members) != 2 {
```
Comment (ldb) [inline http/courseHandler_test.go]: Maybe pull this out of the switch to make it more readable?

File: http/routes.go [hunk 7]
```diff
@@ -9,16 +9,20 @@ func (a *AppServer) authenticatedRoutes() *httprouter.Router {
...
 	router.GET("/api/v1/courses/:id", a.GetCourseHandler())
+	router.GET("/api/v1/courses/:id/users", a.GetMembersHandler())
+	router.POST("/api/v1/courses/:id/users/subscribe", a.AddMembersHandler())
+	router.POST("/api/v1/courses/:id/users/unsubscribe", a.RemoveMembersHandler())
 	router.GET("/api/v1/courses", a.GetAllCoursesHandler())
 
 	// CourseEntries
-	router.POST("/api/v1/courses/:courseID/entries", a.PostCourseEntryHandler())
-	router.PUT("/api/v1/courses/:courseID/entries/:entryID", a.PutCourseEntryHandler())
-	router.DELETE("/api/v1/courses/:courseID/entries/:entryID", a.DeleteCourseEntryHandler())
```
Comment (ldb) [inline http/routes.go]: I think leaving `courseID` here is better to know what ID one is referring to in the respective handlers.

File: service/courseService/courseService.go [hunk 8]
```diff
@@ -37,3 +37,29 @@ func (cS CourseService) GetCourse(id string, cef eduboard.CourseEntryManyFinder)
...
 }
+
+func (cS CourseService) GetCoursesByMember(id string) (error, []eduboard.Course) {
+	return cS.CR.FindByMember(id)
+}
+
+func (cS CourseService) GetMembers(id string, uF eduboard.UserFinder) (error, []eduboard.User) {
+	err, course := cS.CR.FindOneByID(id)
+	if err != nil {
+		return err, []eduboard.User{}
```
Comment (ldb) [inline service/courseService/courseService.go]: Please wrap errors with some additional context here. You can see `courseService` or `courseEntryService` for some examples.

File: service/courseService/courseService_test.go [hunk 9]
```diff
@@ -123,3 +123,201 @@ func TestCourseService_GetCourse(t *testing.T) {
...
+		switch member {
+		case "1":
+			return nil, []eduboard.Course{course1}
+		default:
+			return errors.New("Error fetching courses"), []eduboard.Course{}
+		}
+	}
+
+	for _, v := range testCases {
+		t.Run(fmt.Sprintf("GetCoursesByMember %s", v.name), func(t *testing.T) {
```
Comment (ldb) [inline service/courseService/courseService_test.go]: `go test -v` already prints the name of the testing function, in other cases this would not be printed anyways.

File: service/userService/userService_test.go [hunk 10]
```diff
@@ -197,6 +208,71 @@ func TestUserService_Logout(t *testing.T) {
...
 }
 
+func TestUserService_GetMyCourses(t *testing.T) {
+	t.Parallel()
+	var mockUserRepo mock.UserRepository
+	var mockCourseRepo mock.CourseRepository
+
+	service := UserService{r: &mockUserRepo}
+
+	//	user1 := eduboard.User{ID: "1", Name: "User", Email: "user1@example.com"}
```
Comment (ldb) [inline service/userService/userService_test.go]: Leftover?

**Review by ldb — CHANGES_REQUESTED** [review]
Very good changes overall, especially removing all the `*eduboard.User` returns 👍 
Only a few things left, repository interfaces being the most important one.

File: course.go (hunk 2 above)
Comment (phiphi282) [reply course.go]: The downside with update is, that we need to use the mgo package inside of the service layer.
Is there a way to make the interface smaller without using bson.M in the service layer?

Comment (ldb) [reply course.go]: I was thinking about that. Conclusion was that Go actually has no package hierarchy. So `mgo` is not really the same as `bson` (eventhough they are shipped together). I had the same conflict of mind when adding `bson.ObjectID`s to the domain models. In the end I think its not so bad, that services are defining queries and repositories making sure they get executed correctly. Another solution to this would be to accept `map[string]interface{}` which is type aliased by `bson.M` for convenience. The repository could cast it to the bson type then.

File: README.md [hunk 11]
```diff
@@ -33,20 +33,159 @@ You can start it over the command line (your container name might be different f
...
+        "name": "Mathias",
+        "surname": "Hertzel", (optional)
+        "email": "mathias.hertzel@example.com",
+        "password": "supersecret"
+    }
+    ```
+
+    ```json
+    {
+        name: "Mathias",
```
Comment (ldb) [inline README.md]: Missing Quotes.

File: course.go [hunk 12]
```diff
@@ -50,11 +42,9 @@ type CourseFindUpdater interface {
 }
 
 type CourseRepository interface {
-	CourseInserter
```
Comment (ldb) [inline course.go:41]: Why did you remove this one and not `CourseMemberAdder` and `CourseMemberRemover`?
